File: backend/internal/models/tournament.go
package models

import "time"

// Competition represents a named competition (e.g. "NCAA Tournament")
type Competition struct {
	ID   string `json:"id"`
	Name string `json:"name"`
}

// Season represents a year/season
type Season struct {
	ID   string `json:"id"`
	Year int    `json:"year"`
}

// Tournament represents a basketball tournament in the real world
type Tournament struct {
	ID                   string     `json:"id"`
	Name                 string     `json:"name"`
	Rounds               int        `json:"rounds"` // Total number of rounds in the tournament
	FinalFourTopLeft     string     `json:"finalFourTopLeft"`
	FinalFourBottomLeft  string     `json:"finalFourBottomLeft"`
	FinalFourTopRight    string     `json:"finalFourTopRight"`
	FinalFourBottomRight string     `json:"finalFourBottomRight"`
	StartingAt           *time.Time `json:"startingAt,omitempty"`
	CreatedAt            time.Time  `json:"createdAt"`
	UpdatedAt            time.Time  `json:"updatedAt"`
	DeletedAt            *time.Time `json:"deletedAt,omitempty"`
}

func (t *Tournament) HasStarted(now time.Time) bool {
	if t == nil || t.StartingAt == nil {
		return false
	}
	return !now.Before(*t.StartingAt)
}

const (
	TournamentEditDeniedReasonTournamentMissing = "tournament_missing"
	TournamentEditDeniedReasonTournamentStarted = "tournament_started"
)

func (t *Tournament) CanEditBids(now time.Time, isAdmin bool) (bool, string) {
	if t == nil {
		return false, TournamentEditDeniedReasonTournamentMissing
	}
	if isAdmin {
		return true, ""
	}
	if !t.HasStarted(now) {
		return true, ""
	}
	return false, TournamentEditDeniedReasonTournamentStarted
}

File: frontend/src/types/tournament.ts
export interface Tournament {
  id: string;
  name: string;
  rounds: number;
  winner?: string;
  finalFourTopLeft?: string;
  finalFourBottomLeft?: string;
  finalFourTopRight?: string;
  finalFourBottomRight?: string;
  startingAt?: string;
  createdAt: string;
  updatedAt: string;
}

export interface KenPomStats {
  netRtg: number;
  oRtg: number;
  dRtg: number;
  adjT: number;
}

export interface TournamentTeam {
  id: string;
  schoolId: string;
  tournamentId: string;
  seed: number;
  region: string;
  byes: number;
  wins: number;
  eliminated: boolean;
  createdAt: string;
  updatedAt: string;
  school?: { id: string; name: string };
  kenPom?: KenPomStats;
}

export interface Competition {
  id: string;
  name: string;
}

export interface Season {
  id: string;
  year: number;
}

export interface TournamentModerator {
  id: string;
  email: string;
  firstName: string;
  lastName: string;
}
